README.md
# npme-auth-github
GitHub authentication and authorization strategy for npm Enterprise.

## Implementing your own authentication strategy

### npm Enterprise authentication flow

1. npm CLI calls the frontdoor host with a request to authenticate when `npm login`
is invoked.
2. Frontdoor host calls the authentication webservice with a payload including
the credentials user input.
3. Authentication webservice calls its configured authentication strategy, passing
the payload received to it.
4. Authentication strategy either returns an object with a token and an user
object to persist into the session store if authentication succeeded or no
token and an optional message if authentication failed (see below for the
exact API).
5. If authentication succeeded, frontdoor passes the token back to npm CLI.
npm CLI will further use the token to authorize its request.

### Authentication strategy API

Your module needs to export an `Authenticator` property.  `Authenticator` is
called with an object containing options for the running npm Enterprise instance
and needs to return an object with an `authenticate` function, for example:

```js
var Authenticator = exports.Authenticator = function(opts) {
};

Authenticator.prototype.authenticate = function(credentials, cb) {
};
```

or

```js
exports.Authenticator = function (opts) {
  function authenticate(credentials, cb) {
  }

  return {
    authenticate: authenticate
  };
};
```

The `authenticate` function is called with an object containing a `body` property,
which is what the npm CLI called the frontdoor host with:

```js
{
  body: {
    name: 'foobar',
    email: 'foobar@mycorp.com',
    password: 'iloveicecream'
  }
}
```

Basing on this, you should authenticate the user.

If authentication succeeds, you should call the callback with no error and an
object with `token` and `user` properties. `token` will be passed back to npm
CLI in order to authorize further requests, and `user` will be persisted into
the session store.

```js
cb(null, {
  token: "username-authtoken",
  user: {
    name: 'foobar',
    email: 'foobar@mycorp.com'
  }
});
```

If authentication isn't correct, you should call the callback with no error
and an object with optional `message` property which will be displayed to the
user:

```js
cb(null, {
  message: "You don't work here anymore"
});
```

If authentication errors out (for example, your internal authentication server
is down, you should call the callback with an error object:

```js
cb(new Error('Internal authentication server unreachable'));
```

So, a basic implementation of an authentication strategy based on an abstract
HTTP authentication service could look something like that:

```js
var request = require('request');

var Authenticator = exports.Authenticator = function(opts) {
  this.myAuthenticationHost = opts.myAuthenticationHost;
};

Authenticator.prototype.authenticate = function(credentials, cb) {
  request({
    url: this.myAuthenticationHost + '/auth',
    method: 'POST',
    json: true,
    body: {
      username: credentials.body.username,
      password: credentials.body.password
    }
  }, function (err, res, body) {
    if (err) {
      return cb(err);
    }

    if (res.statusCode !== 200) {
      return cb(null, {
        message: 'Authentication failed'
      });
    }

    return cb(null, {
      token: body.token,
      user: body.user
    });
  });
};
```

## Implementing your own authorization strategy

### npm Enterprise authorization flow

1. With each request npm CLI makes it sends a token previously received from the
frontdoor by the way of `npm login`.
2. Unless specified otherwise, frontdoor checks the authorization token on
every request by calling the authentication webservice.
3. Authentication webservice calls its configured authorization strategy, passing
the token to it and request context to it.
4. Authorization strategy either returns whether authorization was successful
or not.
5. If authorization succeeded, and session store has a session for this token,
frontdoor allows the request to go through.

### Authorization strategy API

Your module needs to export an `Authorizer` property. `Authorizer` is called
with an object containing options for the running npm Enterprise instance
and needs to return an object with an `authorize` function, for example:

```js
var Authorizer = exports.Authorizer = function(opts) {
};

Authorizer.prototype.authorize = function(credentials, cb) {
};
```

or

```js
exports.Authorizer = function (opts) {
  function authorize(credentials, cb) {
  }

  return {
    authorize: authorize
  };
};
```

The `authorize` function is called with an object that looks similar to this:

```js
{
  path: '/mymodule',
  method: 'PUT',
  headers: {
    // ...
    referer: 'npm publish',
    authorization: 'Bearer ...'
  },
  body: {
    // request body
  }
}
```

Basing on this, you should authorize the user.

If authorization succeeds, you should call the callback with no error and `true`.

```js
cb(null, true);
```

If authorization fails, you should call the callback with no error and `false`.

```js
cb(null, false);
```

If authorization errors out (for example, your internal authorization server
is down), you should call the callback with an error object:

```js
cb(new Error('Internal authorization server unreachable'));
```

So, here's an example of using an abstract HTTP authorization service:

```js
var request = require('request');

var Authorizer = exports.Authorizer = function(opts) {
  this.myAuthorizationHost = opts.myAuthorizationHost;
};

Authorizer.prototype.authorize = function(credentials, cb) {
  if (!credentials.headers.authorization ||
      !credentials.headers.authorization.match(/Bearer /)) {
    return cb(null, false);
  }

  request({
    url: this.myAuthorizationHost + '/authorize',
    method: 'POST',
    json: true,
    body: {
      token: credentials.headers.authorization.replace('Bearer ', '')
    }
  }, function (err, res, body) {
    if (err) {
      return cb(err);
    }

    if (res.statusCode !== 200) {
      return cb(null, false);
    }

    return cb(null, true);
  });
};
```

Since you have access to the `package.json`, if one is being sent by npm,
your authorization can involve various checks based on it. For example, GitHub
authorization plugin uses the `repository` field in connection with using GitHub
token as the authorization token to determine if user has write access to the
package they are trying to publish.

## Implementing your own session handler

Session handler is used by the authentication webservice to persist user's name
and email, keyed by the token created by the authentication strategy.

The point of creating custom session handlers is to provide strategy-specific
fallbacks for when user tries authorizing requests against an npm Enterprise
instance without the token present in the session store.  
For example, GitHub strategy provides a custom session store which first checks
Redis for the token, and if it's not present there, it attempts to authenticate
with GitHub with the same token. If GitHub authentication succeeds, user's
GitHub email and login are persisted back to Redis.

### npm Enterprise session flow

#### Authentication

See [npm Enterprise authentication flow](#npm-enterprise-authentication-flow)
for exact description of the authentication flow.

1. Frontdoor host calls the authentication webservice with a payload including
the credentials user input.
2. Authentication webservice calls its configured authentication strategy, passing
the payload received to it.
3. If authentication strategy succeeds, a token is returned.
4. Session handler is called with a key (in form of `"user-" + token`), and
session data to persist.

#### Authorization

See [npm Enterprise authorization flow](#npm-enterprise-authorization-flow)
for exact description of the authorization flow.

1. With each request npm CLI makes it sends a token previously received from the
frontdoor by the way of `npm login`.
2. Unless specified otherwise, frontdoor checks the authorization token on
every request by calling the authentication webservice.
3. Authentication webservice calls its configured authorization strategy, passing
the token and request context to it.
4. Authorization strategy returns whether authorization was successful or not.
5. If authorization succeeded, session store is called with a key (in form of
`"user-" + token`) to retrieve.

### Session handler API
Your module needs to export a `Session` property. `Session` is called with an
object containing options for the running npm Enterprise instance and needs to
return an object with `get` and `set` functions.

```js
var Session = exports.Session = function(opts) {
};

Session.prototype.get = function(key, cb) {
};

Session.prototype.set = function(key, session, cb) {
};
```

or

```js
exports.Session = function (opts) {
  function get(key, cb) {
  }

  function set(key, session, cb) {
  }

  return {
    get: get,
    set: set
  };
};
```

The `get` function is called with a key to retrieve from the session store.
If getting the key from session store succeeds, you should call the callback
with the session content.

```js
cb(null, {
  name: 'foobar',
  email: 'foobar@mycorp.com'
});
```

If getting the key fails, you should call the callback with an error.

```js
cb(new Error('No such key'));
```

The `set` function is called with a key and data to persist into the session
store. If storing the key succeeds, you should call the callback without an
error. If storing the key fails, you should call the callback with an error.

Here's an example of using a Redis-based session store, which fails over to
an abstract HTTP-based authorization service:

```js
var request = require('request');
var redis = require('redis');

var Session = exports.Session = function(opts) {
  this.myAuthorizationHost = opts.myAuthorizationHost;
  this.redis = redis.createClient();
};

Session.prototype._lookup = function(key, cb) {
  var self = this;

  request({
    url: this.myAuthenticationHost + '/authorize',
    method: 'POST',
    json: true,
    body: {
      token: key.split('-').splice(1).join('-')
    }
  }, function (err, res, body) {
    if (err) {
      return cb(err);
    }

    if (res.statusCode !== 200) {
      return cb(new Error('Authorization failed'));
    }

    var session = {
      name: body.username,
      email: body.email
    };

    self.set(key, session, function (err) {
      if (err) {
        return cb(err);
      }
      cb(null, session);
    });
  });
};

Session.prototype.get = function(key, cb) {
  var self = this;
  this.client.get(key, function(err, data) {
    if (err) return cb(err);
    if (data) return cb(null, JSON.parse(data));
    self._lookup(key, cb);
  });
};

Session.prototype.set = function(key, session, cb) {
  this.client.set(key, JSON.stringify(session), cb);
};
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.5.3"></a>
## [1.5.3](https://github.com/npm/npme-auth-github/compare/v1.5.2...v1.5.3) (2017-07-12)


### Bug Fixes

* [@npm](https://github.com/npm)/enterprise-configurator should be a dependency not a dev-dependency ([04f2ef9](https://github.com/npm/npme-auth-github/commit/04f2ef9))



<a name="1.5.2"></a>
## [1.5.2](https://github.com/npm/npme-auth-github/compare/v1.5.1...v1.5.2) (2017-07-11)


### Bug Fixes

* use the version of enterprise-configurator published to the registry ([#41](https://github.com/npm/npme-auth-github/issues/41)) ([bd1b72c](https://github.com/npm/npme-auth-github/commit/bd1b72c))



<a name="1.5.1"></a>
## [1.5.1](https://github.com/npm/npme-auth-github/compare/v1.5.0...v1.5.1) (2017-03-16)


### Bug Fixes

* **authorizer:** Fix organization check when multi-org is configured ([#40](https://github.com/npm/npme-auth-github/issues/40)) ([f5c3954](https://github.com/npm/npme-auth-github/commit/f5c3954))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/npm/npme-auth-github/compare/v1.2.3...v1.5.0) (2017-03-15)


### Features

* "npm On-Site authentication" → "npm Enterprise login" ([#32](https://github.com/npm/npme-auth-github/issues/32)) ([5be278d](https://github.com/npm/npme-auth-github/commit/5be278d))
* add support for multiple GitHub organizations ([#39](https://github.com/npm/npme-auth-github/issues/39)) ([5f0c90e](https://github.com/npm/npme-auth-github/commit/5f0c90e))
* added support for github redirects ([#37](https://github.com/npm/npme-auth-github/issues/37)) ([ea71843](https://github.com/npm/npme-auth-github/commit/ea71843))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/npm/npme-auth-github/compare/v1.3.0...v1.4.0) (2016-05-19)


### Features

* "npm On-Site authentication" → "npm Enterprise login" ([#32](https://github.com/npm/npme-auth-github/issues/32)) ([5be278d](https://github.com/npm/npme-auth-github/commit/5be278d))


test/session-test.js
var lab = require('lab'),
  Lab = exports.lab = lab.script(),
  Code = require('code'),
  nock = require('nock'),
  fs = require('fs'),
  SessionGithub = require('../session.js'),
  redis = require('redis'),
  config = require('@npm/enterprise-configurator').Config({
    githubHost: "https://api.github.com",
    verbose: true
  }),
  client = redis.createClient(),
  token = 'user-foobar';

Lab.experiment('get', function() {

  Lab.beforeEach(function(done) {
    client.del(token, function() {
      done();
    });
  });

  Lab.test('if user is not found in DB, user is looked up via GitHub API', function(done) {
    var session = new SessionGithub({
      githubHost: 'https://github.example.com',
      debug: false
    });

    var githubApi = nock('https://github.example.com')
      .get('/api/v3/user?access_token=foobar')
      .reply(200, fs.readFileSync('./test/fixtures/user-get.json'), {
        'content-type': 'application/json; charset=utf-8'
      });

    session.get(token, function(err, data) {
      Code.expect(data.name).to.deep.equal('bcoe');
      Code.expect(data.email).to.deep.equal('ben@npmjs.com');
      githubApi.done();
      done();
    });
  });

  Lab.test('it creates session in Redis if user found in GitHub API', function(done) {
    var session = new SessionGithub({
      githubHost: 'https://github.example.com',
      debug: false
    });

    var githubApi = nock('https://github.example.com')
      .get('/api/v3/user?access_token=foobar')
      .reply(200, fs.readFileSync('./test/fixtures/user-get.json'), {
        'content-type': 'application/json; charset=utf-8'
      });

    session.get(token, function(err, data) {
      githubApi.done();
      client.get(token, function(err, sessionJson) {
        var session = JSON.parse(sessionJson);
        Code.expect(session.name).to.deep.equal('bcoe');
        Code.expect(session.email).to.deep.equal('ben@npmjs.com');
        done();
      });
    });
  });

  Lab.test('it returns an error if GitHub API returns non 200', function(done) {
    var session = new SessionGithub({
      githubHost: 'https://github.example.com',
      debug: false
    });

    var githubApi = nock('https://github.example.com')
      .get('/api/v3/user?access_token=foobar')
      .reply(500);

    session.get(token, function(err, data) {
      githubApi.done();
      Code.expect(err.code).to.deep.equal(500);
      done();
    });
  });

});


test/authenticate-test.js

var config = require('@npm/enterprise-configurator').Config({
    headless: true
  }),
  lab = require('lab'),
  Lab = exports.lab = lab.script(),
  Code = require('code'),
  AuthenticateGithub = require('../authenticator.js'),
  nock = require('nock'),
  fs = require('fs');

Lab.experiment('getAuthorizationToken', function() {
  Lab.it("returns authorization token if username and password are valid", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'), {
        'content-type': 'application/json; charset=utf-8'
      });

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').done(function(token) {
      Code.expect(token).to.deep.equal('cc84252fd8061b232beb5e345f33b13d120c236c');
      packageApi.done();
      done();
    });
  });

  Lab.it("uses port parsed from githubHost", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com:4444',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com:4444', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'), {
        'content-type': 'application/json; charset=utf-8'
      });

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').done(function(token) {
      Code.expect(token).to.deep.equal('cc84252fd8061b232beb5e345f33b13d120c236c');
      packageApi.done();
      done();
    });
  });

  Lab.it("raises an exception if 401 is returned", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com')
      .post('/api/v3/authorizations')
      .reply(401);

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').catch(function(err) {
      Code.expect(err.code).to.deep.equal(401);
      packageApi.done();
      done();
    }).done();
  });

  Lab.it("raises an exception if 500 is returned", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com')
      .post('/api/v3/authorizations')
      .reply(500);

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').catch(function(err) {
      Code.expect(err.code).to.deep.equal(500);
      packageApi.done();
      done();
    }).done();
  });

  Lab.it("returns authorization token if username and password are valid, and user is a member of the org", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      githubOrg: 'acme',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'), {
        'content-type': 'application/json; charset=utf-8'
      })
      .get('/api/v3/orgs/acme/members/bcoe-test')
      .reply(204);

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').nodeify(function(err, token) {
      Code.expect(!!err).to.equal(false);
      Code.expect(token).to.deep.equal('cc84252fd8061b232beb5e345f33b13d120c236c');
      packageApi.done();
      done();
    });
  });

  Lab.it("returns authorization token if username and password are valid, and user is a member of at least one org", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      githubOrg: 'org1, acme, org3',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'), {
        'content-type': 'application/json; charset=utf-8'
      })
      .get('/api/v3/orgs/acme/members/bcoe-test')
      .reply(204);

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').nodeify(function(err, token) {
      Code.expect(!!err).to.equal(false);
      Code.expect(token).to.deep.equal('cc84252fd8061b232beb5e345f33b13d120c236c');
      packageApi.done();
      done();
    });
  });

  Lab.it("executes callback with an error if user is not a member of the org", function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      githubOrg: 'acme',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'))
      .get('/api/v3/orgs/acme/members/bcoe-test')
      .reply(404);

    authenticateGithub.getAuthorizationToken('bcoe-test', 'foobar').nodeify(function(err, token) {
      Code.expect(err.code).to.equal(401);
      Code.expect(!!token).to.equal(false);
      packageApi.done();
      done();
    });
  });
});

Lab.experiment('authenticate', function() {
  Lab.it('executes callback with token, if successful', function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com', {
        // we should populate the auth headers with appropriate
        // username and password.
        reqheaders: {
          'authorization': 'Basic YmNvZS10ZXN0OmZvb2Jhcg=='
        }
      })
      .post('/api/v3/authorizations', {
        scopes: ["user","public_repo","repo","repo:status","gist"],
        note: 'npm Enterprise login (0)',
        note_url: 'https://www.npmjs.org'
      })
      .reply(200, fs.readFileSync('./test/fixtures/authenticate-success.json'), {
        'content-type': 'application/json; charset=utf-8'
      });

    authenticateGithub.authenticate({
      body: {
        name: 'bcoe-test',
        password: 'foobar'
      }
    }, function(err, res) {
      Code.expect(res.token).to.deep.equal('cc84252fd8061b232beb5e345f33b13d120c236c');
      Code.expect(res.user.name).to.deep.equal('bcoe-test');
      // email should have a sane default, if we fail to look it up.
      Code.expect(res.user.email).to.deep.equal('npme@example.com');
      done();
    });
  });

  Lab.it('executes callback with error if GHE API fails to generate token', function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com')
      .post('/api/v3/authorizations')
      .reply(401);

    authenticateGithub.authenticate({
      body: {
        name: 'bcoe-test',
        password: 'foobar'
      }
    }, function(err, resp) {
      Code.expect(resp.message).to.deep.equal('unauthorized');
      packageApi.done();
      done();
    });
  });

  Lab.it('executes callback with error if password is missing', function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    authenticateGithub.authenticate({
      body: {
        name: 'bcoe-test',
      }
    }, function(err) {
      Code.expect(err.message).to.deep.equal('invalid credentials format');
      done();
    });
  });

  Lab.it('executes callback with error credentials are missing', function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    authenticateGithub.authenticate(null, function(err) {
      Code.expect(err.message).to.deep.equal('invalid credentials format');
      done();
    });
  });

  Lab.it('executes callback with error if GHE API fails with unexpected status code', function(done) {
    var authenticateGithub = new AuthenticateGithub({
      githubHost: 'https://github.example.com',
      timestamp: function() {
        return 0;
      },
      debug: false
    });

    var packageApi = nock('https://github.example.com')
      .post('/api/v3/authorizations')
      .reply(432);

    authenticateGithub.authenticate({
      body: {
        name: 'bcoe-test',
        password: 'foobar'
      }
    }, function(err, resp) {
      Code.expect(err.code).to.deep.equal(432);
      packageApi.done();
      done();
    });
  });
});


session.js

var _ = require('lodash'),
  redis = require('redis'),
  createGithubApi = require('./create-github-api.js');

// Session backed by Redis, but that
// attempts to lookup against GHE if
// the token is not found in the DB.
function SessionGithub(opts) {
  _.extend(this, {
    client: redis.createClient(process.env.LOGIN_CACHE_REDIS),
    githubHost: 'api.github.com',
    debug: true,
    githubPathPrefix: '/api/v3'
  }, require('@npm/enterprise-configurator').Config(), opts)
}

SessionGithub.prototype.get = function(key, cb) {
  var _this = this;

  // First check if we have the session in Redis.
  this.client.get(key, function(err, data) {
    if (err) cb(err);
    // If we do, simply return the existing session.
    else if (data) cb(null, JSON.parse(data));
    else {
      // If we don't, talk to GHE instance. This can happen when, for example,
      // user tries logging into a different npm Enterprise instance with an
      // existing, logged in .npmrc.
      _this._githubLookup(key, cb);
    }
  });
};

SessionGithub.prototype._githubLookup = function(key, cb) {
  var _this = this,
    github = createGithubApi(this),
    // extract the GHE key from the user-<token> string.
    token = key.split('-').splice(1).join('-');

  // the token authenticator creates is a GitHub OAuth token, so we can simply
  // try authenticating with it and then fetching the authenticated user.
  github.authenticate({
    type: 'oauth',
    token: token
  });

  github.users.get({}, function(err, res) {
    if (err) cb(err);
    else if (res.code < 200 || res.code >= 400) cb(Error('status = ' + res.code));
    else {
      var session = {
        name: res.login,
        email: res.email || 'npme@example.com'
      };

      // If authentication with GitHub succeeded, persist the GitHub login and
      // email in our session store.
      _this.set(key, session, function(err) {
        if (err) cb(err);
        else cb(undefined, session);
      });
    }
  });
};

SessionGithub.prototype.set = function(key, session, cb) {
  this.client.set(key, JSON.stringify(session), cb);
};

module.exports = SessionGithub;


create-github-api.js
var _ = require('lodash'),
  GithubApi = require('github'),
  config = require('@npm/enterprise-configurator').Config(),
  url = require('url');

// helper for generating a GitHub API instance
// from a GitHub or GitHub Enterprise URL.
module.exports = function(githubParams) {
  var parsedUrl = url.parse(githubParams.githubHost),
    protocol = parsedUrl.protocol ? parsedUrl.protocol.replace(':', '') : 'https',
    githubOpts = {
      version: "3.0.0",
      debug: githubParams.debug,
      timeout: 30000,
      tokenName: 'npm Enterprise solution'
    };

  // GHE Domain.
  if (parsedUrl.host.indexOf('api.github.com') === -1) {
    githubOpts.protocol = protocol;
    githubOpts.port = parsedUrl.port ? parsedUrl.port : {'http': 80, 'https': 443}[protocol];
    githubOpts.host = parsedUrl.host.replace(/:.*$/, '');
    githubOpts.pathPrefix = githubParams.githubPathPrefix;
  }

  return new GithubApi(githubOpts);
};


authenticator.js
var parseUrl = require('url'),
  logger = require('@npm/enterprise-configurator').logger(),
  Promise = require('bluebird'),
  _ = require('lodash'),
  createGithubApi = require('./create-github-api.js'),
  config = require('@npm/enterprise-configurator').Config();

function AuthenticateGithub(opts) {
  _.extend(this, {
    packagePath: null, // required, name-spaced package name.
    debug: false,
    githubHost: config.githubHost,
    githubPathPrefix: '/api/v3',
    githubOrg: config.githubOrg,
    // label the token that we generate.
    note: 'npm Enterprise login',
    noteUrl: 'https://www.npmjs.org'
  }, opts);
}

// Reach out to GitHub API to authenticate the user. Create an authorization
// token and pass it back as it is, allowing the authorizer to verify it through
// GitHub API later on.
AuthenticateGithub.prototype.authenticate = function(credentials, cb) {
  if (!this._validateCredentials(credentials)) return cb(Error('invalid credentials format'));

  var body = credentials.body,
    username = body.name,
    password = body.password,
    twoFactorCode = body.code;

  this.getAuthorizationToken(username, password, twoFactorCode)
    .then(function(token) {
      return {
        token: token,
        user: {
          name: username,
          email: body.email || 'npme@example.com'
        }
      };
    }, function(err) {
      if (err.code === 401) {
        err.message = 'unauthorized';
        // this is a failure to auth, but not an error
        return err;
      } else if (err.code === 500) {
        err.message = 'GitHub enterprise unavailable';
      }
      // this is an error state
      throw err;
    })
    .nodeify(cb);
};

AuthenticateGithub.prototype._validateCredentials = function(credentials) {
  if (!credentials) return false;
  if (!credentials.body) return false;
  if (!credentials.body.name || !credentials.body.password) return false;
  return true;
};

// Actually create the authorization token.
AuthenticateGithub.prototype.getAuthorizationToken = function(username, password, twoFactorCode) {
  var _this = this,
    github = createGithubApi(this),
    headers = {};

  github.authenticate({
    type: 'basic',
    username: username,
    password: password
  });

  if (twoFactorCode) headers['X-GitHub-OTP'] = twoFactorCode;

  return new Promise(function(resolve, reject) {
    github.authorization.create({
      scopes: ["user", "public_repo", "repo", "repo:status", "gist"],
      // timestamp helps prevent duplicate tokens.
      note: _this.note + ' (' + _this.timestamp() + ')',
      note_url: _this.noteUrl,
      headers: headers
    }, function(err, res) {
      if (err) reject(err);
      else resolve(res.token);
    });
  }).then(this.githubOrg && function(token) {
    var orgs = Array.isArray(_this.githubOrg) ? _this.githubOrg : _this.githubOrg.split(/,[\s]*/);
    return Promise.any([].concat(orgs).map(function (githubOrg) {
      return new Promise(function(resolve, reject) {
         github.orgs.checkMembership({ user: username, org: githubOrg }, function(err, res) {
           if (err) reject(err);
           else resolve(token);
         });
      });
    })).catch(function(err) {
      // handle possible AggregateError, which is a collection of promise errors
      err = err instanceof Promise.AggregateError ? err[0] : err;

      if (err.code === 404) {
        err.code = 401;
        err.message = 'unauthorized';
      } else if (err.code === 500) {
        err.message = 'GitHub enterprise unavailable';
      }
      // this is an error state
      throw err;
    });
  });
};

AuthenticateGithub.prototype.timestamp = function() {
  return new Date().getTime();
};

module.exports = AuthenticateGithub;
